### PR TITLE
[Feature](func) Support the third argument for regexp_extract_all and regexp_extract_all_array

### PR DESCRIPTION
**Pull request by ccl125**
### What problem does this PR solve?

Issue Number: related #48203 (Spark: `regexp_extract_all` for the third argument)

Problem Summary:

`regexp_extract_all` / `regexp_extract_all_array` only accepted 2 arguments and always extracted the first capturing group. This PR adds the optional third `group` argument, following Spark semantics:

- `regexp_extract_all(str, pattern)` — unchanged, extracts group 1
- `regexp_extract_all(str, pattern, 0)` — extracts the whole match (also works for patterns without capturing groups)
- `regexp_extract_all(str, pattern, N)` — extracts capturing group N
- group index outside `[0, number_of_capturing_groups]` → error (same as Spark); non-participating groups yield empty strings

Implementation:
- FE: `RegexpExtractAll` / `RegexpExtractAllArray` gain 3-arg signatures; the 2-arg form stays a 2-child node (no default padding) so old-FE/new-BE combinations keep working during rolling upgrades
- BE: both arities are registered as separate variadic forms (mirroring `regexp_replace`'s ThreeParamTypes/FourParamTypes); `RegexpExtractEngine::match_all_and_extract` takes a group index; illegal indexes (outside `[0, number_of_capturing_groups]`) raise `INVALID_ARGUMENT`, and non-participating groups contribute empty strings — both consistent with Spark

### Release note

Support the third argument (group index) for `regexp_extract_all` and `regexp_extract_all_array`.

### Check List (For Author)

- Test
    - [x] Regression test
    - [x] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:

  New BE unit cases in `function_like_test.cpp` cover group 0 / other groups / out-of-range / negative index for both functions; existing 2-arg cases were ported to the normalized 3-arg form with unchanged expectations. New regression cases added to `test_string_function_regexp`.

- Behavior changed:
    - [x] No.
    - [ ] Yes.

  2-argument calls keep the exact previous behavior (group 1 extraction, empty result when the pattern has no capturing group).

- Does this need documentation?
    - [ ] No.
    - [x] Yes. https://github.com/apache/doris-website/pull/4022

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label

### CODE REVIEW

**Comment by hello-stephen** [comment]
Thank you for your contribution to Apache Doris.
Don't know what should be done next? See [How to process your PR](https://cwiki.apache.org/confluence/display/DORIS/How+to+process+your+PR).

Please clearly describe your PR:
1. What problem was fixed (it's best to include specific error reporting information). How it was fixed.
2. Which behaviors were modified. What was the previous behavior, what is it now, why was it modified, and what possible impacts might there be.
3. What features were added. Why was this function added?
4. Which code was refactored and why was this part of the code refactored?
5. Which functions were optimized and what is the difference before and after the optimization?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/RegexpExtractAll.java [hunk 1]
```diff
@@ -61,7 +74,7 @@ private RegexpExtractAll(ScalarFunctionParams functionParams) {
      */
     @Override
     public RegexpExtractAll withChildren(List<Expression> children) {
-        Preconditions.checkArgument(children.size() == 2);
+        Preconditions.checkArgument(children.size() == 2 || children.size() == 3);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/RegexpExtractAll.java:76]:
```suggestion
        Preconditions.checkArgument(children.size() == 3);
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/RegexpExtractAllArray.java [hunk 2]
```diff
@@ -64,7 +77,7 @@ private RegexpExtractAllArray(ScalarFunctionParams functionParams) {
      */
     @Override
     public RegexpExtractAllArray withChildren(List<Expression> children) {
-        Preconditions.checkArgument(children.size() == 2);
+        Preconditions.checkArgument(children.size() == 2 || children.size() == 3);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/RegexpExtractAllArray.java:79]:
```suggestion
        Preconditions.checkArgument(children.size() == 3);
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/RegexpExtractAll.java (hunk 1 above)
Comment (ccl125) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/RegexpExtractAll.java:76]: Thanks for the review! Fixed in aa701aa.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/RegexpExtractAllArray.java (hunk 2 above)
Comment (ccl125) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/RegexpExtractAllArray.java:79]: Thanks for the review! Fixed in aa701aa.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/RegexpExtractAllArray.java [hunk 3]
```diff
@@ -38,20 +39,32 @@
...
-        implements BinaryExpression, ExplicitlyCastableSignature, AlwaysNullable, PropagateNullLiteral {
+        implements ExplicitlyCastableSignature, AlwaysNullable, PropagateNullLiteral {
 
     public static final List<FunctionSignature> SIGNATURES = ImmutableList.of(
             FunctionSignature.ret(ArrayType.of(VarcharType.SYSTEM_DEFAULT))
                     .args(VarcharType.SYSTEM_DEFAULT, VarcharType.SYSTEM_DEFAULT),
             FunctionSignature.ret(ArrayType.of(StringType.INSTANCE))
-                    .args(StringType.INSTANCE, StringType.INSTANCE)
+                    .args(StringType.INSTANCE, StringType.INSTANCE),
+            FunctionSignature.ret(ArrayType.of(VarcharType.SYSTEM_DEFAULT))
```
Comment (linrrzqqq) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/RegexpExtractAllArray.java:48]: It's better not to change the parameter limit from two to a hard limit of three. I think  support variable parameters is better.

Doris's version upgrades are generally rolling upgrades, first upgrading the backend (be) and then the frontend (fe).An old FE sends two arguments to a new BE, while an old FE pads two-argument SQL calls to three arguments and therefore cannot execute them on an new BE.

Could we keep the original two-argument expression shape and make the new BE accept both 2 and 3 arguments? When only two arguments are provided, the BE can simply use group index 1 as the default. This would preserve existing two-argument queries during a rolling upgrade

File: be/src/exprs/function/function_regexp.cpp [hunk 4]
```diff
@@ -136,13 +136,17 @@ struct RegexpExtractEngine {
...
+    void match_all_and_extract(const char* data, size_t size, int index,
                                std::vector<std::string>& results) const {
+        if (index < 0) {
+            return;
+        }
         if (is_re2()) {
             int max_matches = 1 + re2_regex->NumberOfCapturingGroups();
-            if (max_matches < 2) {
-                return; // No capturing groups
+            if (index >= max_matches) {
```
Comment (linrrzqqq) [inline be/src/exprs/function/function_regexp.cpp:218]: The behavior for `idx >= max_matches || index < 0` better refer to Spark throwing an error.

```sql
spark-sql (default)> SELECT regexp_extract_all('100-200, 300-400', '(\\d+)-(\\d+)', -1); 
[INVALID_PARAMETER_VALUE.REGEX_GROUP_INDEX] The value of parameter(s) `idx` in `regexp_extract_all` is invalid: Expects group index between 0 and 2, but got -1. SQLSTATE: 22023
```

File: be/src/exprs/function/function_regexp.cpp [hunk 5]
```diff
@@ -159,9 +163,9 @@ struct RegexpExtractEngine {
                     pos += 1;
                     continue;
                 }
-                // Extract first capturing group
-                if (matches.size() > 1 && !matches[1].empty()) {
-                    results.emplace_back(matches[1].data(), matches[1].size());
+                // Extract the capturing group with the given index
+                if (static_cast<size_t>(index) < matches.size() && !matches[index].empty()) {
```
Comment (linrrzqqq) [inline be/src/exprs/function/function_regexp.cpp]: should not judge `matches[idx].empty()` anymore:
```sql
spark-sql (default)> select regexp_extract_all('a b', '(a)|(b)', 2);
["","b"]
```

File: be/src/exprs/function/function_regexp.cpp [hunk 6]
```diff
@@ -174,8 +178,8 @@ struct RegexpExtractEngine {
             boost::match_results<const char*> matches;
 
             while (boost::regex_search(search_start, search_end, matches, *boost_regex)) {
-                if (matches.size() > 1 && matches[1].matched) {
-                    results.emplace_back(matches[1].str());
+                if (static_cast<size_t>(index) < matches.size() && matches[index].matched) {
```
Comment (linrrzqqq) [inline be/src/exprs/function/function_regexp.cpp]: ditto

File: regression-test/suites/query_p0/sql_functions/string_functions/test_string_function_regexp.groovy [hunk 7]
```diff
@@ -255,4 +255,11 @@ suite("test_string_function_regexp") {
 
     qt_sql_field4 "SELECT FIELD('21','2130', '2131', '21');"
     qt_sql_field5 "SELECT FIELD(21, 2130, 21, 2131);"
+
+    qt_sql_regexp_extract_all_group0 "select regexp_extract_all('x=a3&x=18abc&x=2&y=3&x=4&x=17bcd', 'x=([0-9]+)([a-z]+)', 0);"
+    qt_sql_regexp_extract_all_group2 "select regexp_extract_all('x=a3&x=18abc&x=2&y=3&x=4&x=17bcd', 'x=([0-9]+)([a-z]+)', 2);"
+    qt_sql_regexp_extract_all_group_negative "select regexp_extract_all('abc', '(b)', -1);"
+    qt_sql_regexp_extract_all_array_group0 "select regexp_extract_all_array('x=a3&x=18abc&x=2&y=3&x=4&x=17bcd', 'x=([0-9]+)([a-z]+)', 0);"
+    qt_sql_regexp_extract_all_array_group2 "select regexp_extract_all_array('x=a3&x=18abc&x=2&y=3&x=4&x=17bcd', 'x=([0-9]+)([a-z]+)', 2);"
+    qt_sql_regexp_extract_all_array_group_out_of_range "select regexp_extract_all_array('hitdecisiondlist', '(i)(.*?)(e)', 3);"
```
Comment (linrrzqqq) [inline regression-test/suites/query_p0/sql_functions/string_functions/test_string_function_regexp.groovy]: Please add more comprehensive tests, covering more cases, including but not limited to: including column input, including null, illegal index...

Comment (linrrzqqq) [reply regression-test/suites/query_p0/sql_functions/string_functions/test_string_function_regexp.groovy]: btw, why this case named after `_out_of_range`?

I think `(i)(.*?)(e)` has three capturing groups, so index 3 is valid and should extract "e". The first out-of-range index is 4.

seems to be some issue in ur impl:
```sql
spark-sql (default)> select regexp_extract_all('hitdecisiondlist', '(i)(.*?)(e)', 3);
["e"]
```

**Comment by linrrzqqq** [comment]
also remember to edit doc: https://github.com/apache/doris-website

dev & 4.x versions of zh & en

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/RegexpExtractAll.java (hunk 1 above)
Comment (ccl125) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/RegexpExtractAll.java:76]: Update: the design changed based on the rolling-upgrade feedback below — the FE no longer pads a default index, so a node can legitimately have 2 or 3 children and the assertion stays `2 || 3`. The BE now accepts both arities.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/RegexpExtractAllArray.java (hunk 2 above)
Comment (ccl125) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/RegexpExtractAllArray.java:79]: Same as above — reverted to `2 || 3` since the FE no longer normalizes to 3 arguments.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/RegexpExtractAllArray.java (hunk 3 above)
Comment (ccl125) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/RegexpExtractAllArray.java:48]: Good catch, thanks! Reworked in d387667: the FE no longer pads, and the BE now registers both a 2-arg and a 3-arg variadic form (mirroring regexp_replace's ThreeParamTypes/FourParamTypes), so an old FE keeps working against a new BE during rolling upgrades.

File: be/src/exprs/function/function_regexp.cpp (hunk 4 above)
Comment (ccl125) [reply be/src/exprs/function/function_regexp.cpp:218]: Done — the index contract now matches Spark: an index outside [0, number_of_capturing_groups] fails with InvalidArgument, with the valid range in the message.

File: be/src/exprs/function/function_regexp.cpp (hunk 5 above)
Comment (ccl125) [reply be/src/exprs/function/function_regexp.cpp]: Done — non-participating groups now contribute empty strings instead of being skipped (your `(a)|(b)` group-2 example is covered in the unit tests).

File: be/src/exprs/function/function_regexp.cpp (hunk 6 above)
Comment (ccl125) [reply be/src/exprs/function/function_regexp.cpp]: Done, both the re2 and the boost path keep empty groups now.

File: regression-test/suites/query_p0/sql_functions/string_functions/test_string_function_regexp.groovy (hunk 7 above)
Comment (ccl125) [reply regression-test/suites/query_p0/sql_functions/string_functions/test_string_function_regexp.groovy]: Expanded: column input (three group indexes over the test table), NULL literals for string/pattern/index, illegal indexes via exception blocks, and the non-participating-group case.

Comment (ccl125) [reply regression-test/suites/query_p0/sql_functions/string_functions/test_string_function_regexp.groovy]: You were right, I miscounted — `(i)(.*?)(e)` has three capturing groups and index 3 extracts "e". Fixed the unit and regression cases; index 4 is now covered as the out-of-range error.

**Comment by ccl125** [comment]
Thanks for the thorough scan! All five points are addressed in 1ca8dcf (pushed):

1. **2-arg regression** — the out-of-range error now only applies to the explicit three-argument form; two-argument calls on patterns without capturing groups keep the legacy empty-result behavior (covered by a ported unit case).
2. **Truncated pattern literal** — fixed, the string is 11 bytes.
3. **Dropped test table** — the column-input cases now recreate `test_string_function_regexp` before running.
4. **re2 advancing** — the loop now advances via the match pointer offset instead of a textual `find` (this and the next one were pre-existing issues in `match_all_and_extract`, fixed here since the function was being reworked anyway; regression case added).
5. **boost zero-width advancing** — now advances from `matches[0].first + 1` instead of the old origin.

File: be/src/exprs/function/function_regexp.cpp [hunk 8]
```diff
@@ -159,29 +165,35 @@ struct RegexpExtractEngine {
...
+                        results.emplace_back();
+                    }
                 }
-                // Move position forward
-                auto offset = std::string(str_pos, str_size)
-                                      .find(std::string(matches[0].data(), matches[0].size()));
-                pos += offset + matches[0].size();
+                // Move position forward. matches[0] points into the searched string,
+                // so its address gives the exact match offset — no textual find needed.
+                pos += (matches[0].data() - str_pos) + matches[0].size();
```
[the remark this answers is not in the document]
Comment (ccl125) [reply be/src/exprs/function/function_regexp.cpp]: Fixed in cf47ba1 — the re2 loop now calls `Match` on the original subject with an advancing start position (so `^` and left context refer to the original string), and the boost loop keeps the original begin reachable with `match_prev_avail` and suppresses bol at the advanced origin with `match_not_bol`. Added `('aa', '^a', 0) -> ['a']` as a regression case.

[the remark this answers is not in the document]
Comment (ccl125) [reply be/src/exprs/function/function_regexp.cpp]: Fixed in 4737a9d — switched the boost flags to the documented repeated-search idiom `match_prev_avail | match_not_bob` (with match_prev_avail set, match_not_bol is ignored; match_not_bob keeps `\A` anchored to the start of the original buffer).

**Comment by ccl125** [comment]
Hi @linrrzqqq @morrySnow, all review feedback has been addressed (latest: the boost repeated-search flags in 4737a9d). Could someone help trigger `run buildall` when convenient so we can see the full CI result? Thanks!

**Comment by linrrzqqq** [comment]
run buildall

**Comment by hello-stephen** [comment]
# FE UT Coverage Report
Increment line coverage `  0.00% (0/16)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/66050_4737a9dc3e37552c486c92980202530c6fdd2606/fe_increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/66050_4737a9dc3e37552c486c92980202530c6fdd2606/fe_report/index.html)

**Comment by hello-stephen** [comment]
<details>
<summary>TPC-H: <b>Total hot run time: 17000 ms</b></summary>

```
machine: 'aliyun_ecs.c7a.8xlarge_32C64G'
scripts: https://github.com/apache/doris/tree/master/tools/tpch-tools
Tpch sf100 test result on commit 4737a9dc3e37552c486c92980202530c6fdd2606, data reload: false

------ Round 1 ----------------------------------
orders	Doris	NULL	NULL	0	0	0	NULL	0	NULL	NULL	2023-12-26 18:27:23	2023-12-26 18:42:55	NULL	utf-8	NULL	NULL	
============================================
q1	17585	3112	3075	3075
q2	q3	10895	848	517	517
q4	4682	256	203	203
q5	7671	572	384	384
q6	142	115	92	92
q7	525	476	384	384
q8	9251	937	914	914
q9	3553	2383	2360	2360
q10	6510	868	738	738
q11	456	253	237	237
q12	698	408	329	329
q13	17877	1536	1175	1175
q14	153	148	138	138
q15	q16	479	389	375	375
q17	809	787	759	759
q18	3112	2246	2258	2246
q19	1214	931	676	676
q20	680	526	460	460
q21	5564	1708	1878	1708
q22	336	272	230	230
Total cold run time: 92192 ms
Total hot run time: 17000 ms

----- Round 2, with runtime_filter_mode=off -----
orders	Doris	NULL	NULL	150000000	42	6422171781	NULL	22778155	NULL	NULL	2023-12-26 18:27:23	2023-12-26 18:42:55	NULL	utf-8	NULL	NULL	
============================================
q1	3497	3375	3383	3375
q2	q3	2242	2266	2124	2124
q4	1200	1159	887	887
q5	2200	2126	2092	2092
q6	166	117	91	91
q7	1052	901	871	871
q8	1605	1422	1425	1422
q9	3121	3159	3091	3091
q10	1846	1776	1625	1625
q11	368	273	258	258
q12	451	426	353	353
q13	1498	1545	1164	1164
q14	174	163	169	163
q15	q16	400	385	367	367
q17	1068	1031	1034	1031
q18	4967	4382	4728	4382
q19	866	867	852	852
q20	962	952	824	824
q21	3517	3234	3246	3234
q22	408	341	310	310
Total cold run time: 31608 ms
Total hot run time: 28516 ms
```
</details>

**Comment by hello-stephen** [comment]
<details>
<summary>TPC-DS: <b>Total hot run time: 80644 ms</b></summary>

```
machine: 'aliyun_ecs.c7a.8xlarge_32C64G'
scripts: https://github.com/apache/doris/tree/master/tools/tpcds-tools
TPC-DS sf100 test result on commit 4737a9dc3e37552c486c92980202530c6fdd2606, data reload: false

query5	4254	425	355	355
query6	452	174	162	162
query7	4824	471	261	261
query8	333	142	117	117
query9	8681	2924	2911	2911
query10	410	258	218	218
query11	5459	1042	937	937
query12	120	75	75	75
query13	1207	442	334	334
query14	6031	2222	2111	2111
query14_1	2012	1984	1973	1973
query15	194	119	115	115
query16	1010	377	362	362
query17	829	451	370	370
query18	2359	335	243	243
query19	195	146	110	110
query20	72	71	69	69
query21	230	115	101	101
query22	5440	5275	5353	5275
query23	6646	6186	6009	6009
query23_1	6241	6039	6070	6039
query24	7350	1109	815	815
query24_1	781	785	778	778
query25	510	307	260	260
query26	1254	276	164	164
query27	2712	440	288	288
query28	4637	1512	1495	1495
query29	1113	448	356	356
query30	284	182	151	151
query31	856	432	356	356
query32	134	52	50	50
query33	512	227	178	178
query34	1138	838	496	496
query35	413	406	346	346
query36	596	566	528	528
query37	141	81	74	74
query38	1104	844	814	814
query39	493	486	480	480
query39_1	446	470	445	445
query40	338	124	116	116
query41	68	69	68	68
query42	84	80	79	79
query43	250	246	211	211
query44	
query45	112	100	103	100
query46	782	821	519	519
query47	775	751	711	711
query48	299	267	243	243
query49	566	230	188	188
query50	858	335	258	258
query51	8129	8077	8052	8052
query52	85	81	69	69
query53	212	230	161	161
query54	282	210	195	195
query55	109	61	56	56
query56	248	237	215	215
query57	704	676	641	641
query58	309	200	196	196
query59	1236	1220	1095	1095
query60	288	220	189	189
query61	108	114	132	114
query62	487	218	174	174
query63	181	158	153	153
query64	2778	684	601	601
query65	
query66	2028	305	245	245
query67	9860	9901	9930	9901
query68	
query69	393	225	198	198
query70	633	624	633	624
query71	313	252	244	244
query72	2443	1419	1561	1419
query73	771	594	358	358
query74	1650	1246	1162	1162
query75	1273	1157	1055	1055
query76	2341	710	551	551
query77	260	261	215	215
query78	3799	3587	3251	3251
query79	3314	784	576	576
query80	1719	420	359	359
query81	682	205	182	182
query82	1205	138	103	103
query83	338	251	232	232
query84	
query85	954	442	385	385
query86	630	170	175	170
query87	1037	986	875	875
query88	4290	2133	2102	2102
query89	374	227	200	200
query90	2274	149	147	147
query91	163	141	121	121
query92	88	43	45	43
query93	3024	1066	821	821
query94	826	266	218	218
query95	644	450	353	353
query96	850	574	283	283
query97	1051	1071	1029	1029
query98	196	131	130	130
query99	521	337	310	310
Total cold run time: 180158 ms
Total hot run time: 80644 ms
```
</details>

**Comment by hello-stephen** [comment]
# BE Regression && UT Coverage Report
Increment line coverage `100% (0/0)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/66050_4737a9dc3e37552c486c92980202530c6fdd2606_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/66050_4737a9dc3e37552c486c92980202530c6fdd2606_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 75.71% (34090/45029) |
| Line Coverage     | 60.64% (383825/632916) |
| Region Coverage   | 56.83% (322278/567116) |
| Branch Coverage   | 57.65% (146862/254758) |

**Comment by ccl125** [comment]
Thanks for triggering buildall! Everything passed except the FE coverage gate (our FE change is only declarative signatures, so the increment had no UT coverage). Added a dedicated FE unit test for both function classes in 2817282 covering constructors/signatures/withChildren for the 2-arg and 3-arg forms. Could you help trigger `run buildall` once more?

File: be/src/exprs/function/function_regexp.cpp
Comment (linrrzqqq) [inline be/src/exprs/function/function_regexp.cpp:174]: > For the newly supported `idx = 0`, this drops valid zero-length matches before the selected group is appended, and `while (pos < size)` also skips a terminal match. For example, `regexp_extract_all('b', 'a*', 0)` returns `[]` here, while Spark's matcher yields ["", ""] (offsets 0 and 1); empty input/pattern should similarly yield one empty match. Please emit the selected group for successful empty matches and advance safely while allowing one terminal-position search.

fix this

Comment (linrrzqqq) [reply be/src/exprs/function/function_regexp.cpp:174]: also fix : https://github.com/apache/doris/pull/66050#discussion_r3681145559 and https://github.com/apache/doris/pull/66050#discussion_r3681145562

Comment (ccl125) [reply be/src/exprs/function/function_regexp.cpp:174]: Fixed in 1bcc272 — explicit group-index calls now follow Spark exactly: zero-width matches are emitted ('b','a*',0 -> ['',''] and '' -> ['']), one terminal-position search is allowed, and advancement is UTF-8-safe (whole characters, no continuation-byte splits). The legacy two-argument form keeps the old skip behavior, covered by a dedicated case. New unit cases added for all three examples.

Comment (ccl125) [reply be/src/exprs/function/function_regexp.cpp:174]: Also fixed in 1bcc272 — input null maps are now merged before validation, so `(NULL,'(a)',99), ('hitdecisiondlist','(i)(.*?)(e)',3)` yields NULL for the first row and ['e'] for the second instead of aborting on the garbage index. A dedicated two-row nullable-block test covers it. The UTF-8-unsafe advance in the boost path is fixed the same way (whole-character steps via get_utf8_byte_length).

File: be/src/exprs/function/function_regexp.cpp [hunk 9]
```diff
@@ -136,52 +137,93 @@ struct RegexpExtractEngine {
...
             size_t pos = 0;
-            while (pos < size) {
-                const char* str_pos = data + pos;
-                size_t str_size = size - pos;
+            while (pos <= size) {
                 std::vector<re2::StringPiece> matches(max_matches);
-                bool success = re2_regex->Match(re2::StringPiece(str_pos, str_size), 0, str_size,
+                // Search within the original subject starting from pos, so `^` stays
+                // anchored to the beginning of the original string.
+                bool success = re2_regex->Match(re2::StringPiece(data, size), pos, size,
```
[the remark this answers is not in the document]
Comment (ccl125) [reply be/src/exprs/function/function_regexp.cpp:227]: Fixed in ec1c238 — the legacy two-argument form now runs the pre-change substring loop verbatim, so `^` anchoring and suffix behavior are byte-identical to before this PR; only the explicit group-index form searches the original subject.

File: be/src/exprs/function/function_regexp.cpp [hunk 10]
```diff
@@ -136,52 +137,93 @@ struct RegexpExtractEngine {
...
+                        break;
+                    }
                     pos += 1;
                     continue;
                 }
-                // Extract first capturing group
-                if (matches.size() > 1 && !matches[1].empty()) {
-                    results.emplace_back(matches[1].data(), matches[1].size());
+                // Extract the capturing group with the given index
+                if (static_cast<size_t>(index) < matches.size()) {
```
[the remark this answers is not in the document]
Comment (ccl125) [reply be/src/exprs/function/function_regexp.cpp:234]: Fixed in ec1c238 — the legacy loop again appends only the first group when non-empty (the unconditional append now lives only in the Spark loop used by explicit group-index calls).

File: be/src/exprs/function/function_regexp.cpp [hunk 11]
```diff
@@ -136,52 +137,93 @@ struct RegexpExtractEngine {
...
+                    } else {
+                        results.emplace_back();
+                    }
+                }
+                if (whole.empty()) {
+                    if (pos >= size) {
+                        break;
+                    }
+                    // Advance one full UTF-8 character, never into a continuation byte.
+                    pos += get_utf8_byte_length(static_cast<uint8_t>(data[pos]));
```
[the remark this answers is not in the document]
Comment (ccl125) [reply be/src/exprs/function/function_regexp.cpp]: Fixed in ec1c238 — zero-width advancement now starts from the returned match location (`whole.data()` + one UTF-8 character), not from `data[pos]`.

File: be/src/exprs/function/function_regexp.cpp [hunk 12]
```diff
@@ -136,52 +137,93 @@ struct RegexpExtractEngine {
...
-                    if (search_start == search_end) {
+                if (empty_match) {
+                    // Advance past the zero-width match itself (one full UTF-8
+                    // character), otherwise a match found after the origin would be
+                    // emitted twice or the next search could split a multibyte char.
+                    if (matches[0].first == search_end) {
                         break;
                     }
-                    search_start += 1;
+                    search_start =
```
[the remark this answers is not in the document]
Comment (ccl125) [reply be/src/exprs/function/function_regexp.cpp]: Done in ec1c238 — both BE files now pass `clang-format-16 --dry-run --Werror` locally.

File: be/src/exprs/function/function_regexp.cpp [hunk 13]
```diff
@@ -188,6 +195,87 @@ struct RegexpExtractEngine {
...
+                        results.emplace_back();
+                    }
+                }
+                if (whole.empty()) {
+                    if (whole.data() >= data + size) {
+                        break;
+                    }
+                    // Advance one full UTF-8 character FROM THE MATCH LOCATION, never
+                    // into a continuation byte and never rescanning the same match.
+                    pos = (whole.data() - data) +
```
[the remark this answers is not in the document]
Comment (ccl125) [reply be/src/exprs/function/function_regexp.cpp]: Fixed in 8e9b6b7 — advancement after an empty match is now one UTF-16 code unit (Java Matcher semantics): a 4-byte UTF-8 character advances by half, everything shorter by the whole character, so supplementary characters yield the same empty-match count as Spark.

File: be/src/exprs/function/function_regexp.cpp [hunk 14]
```diff
@@ -188,6 +195,87 @@ struct RegexpExtractEngine {
...
+            const char* search_start = data;
+            const char* search_end = data + size;
+            boost::match_results<const char*> matches;
+
+            // Keep the original subject start reachable: match_prev_avail lets
+            // look-behind assertions see characters before search_start, and
+            // match_not_bob keeps `\A` anchored to the start of the original
+            // buffer (Boost's documented repeated-regex_search idiom; with
+            // match_prev_avail set, `^` is decided by the preceding character).
+            while (search_start <= search_end &&
```
[the remark this answers is not in the document]
Comment (ccl125) [reply be/src/exprs/function/function_regexp.cpp:264]: On `\G`: this is a pre-existing limitation rather than a regression — the pre-PR loop also re-searched fresh substrings every iteration, so `\G` never had previous-match state either. Note that RE2 does not support `\G` at all, so such patterns only ever reach the boost path (enable_extended_regex). Making `\G` Spark-exact needs a stateful matcher abstraction this function does not have; I'd rather leave it out of scope here and track it separately if the maintainers want it.

**Comment by ccl125** [comment]
Hi @linrrzqqq, gentle ping — all feedback so far has been addressed: the legacy/Spark loop split, Spark-exact empty-match semantics (UTF-16 code-unit advancement), null-row handling, and the FE coverage test. Could you take another look at the latest commits? And if the code looks good, a `run buildall` would let us confirm everything end to end. Thanks!
